wordpress-mobile/WordPress-iOS
# Issue 19058 - Sort user-mentions suggestions by prefix first, then alphabetically

Closes https://github.com/wordpress-mobile/WordPress-iOS/issues/19058

## Description

In this PR, I'm addressing the sorting change requested. I'm displaying first the users whose names begin with a prefix provided by the user, then alphabetically. The first criteria are to sort alphabetically by users whose usernames or display names begin with the provided prefix. If both username and display name begin with the prefix, the display name will be prioritized in the sorting. 

## Details

- I’m using localizedCaseInsensitiveCompare to compare `displayNames`, because the user can have the displayName in different languages. The comparison is localizable and case-insensitive.

- I’m using `<` to compare usernames because they are very restricted, and the lexicographical comparison should work for them.

- I'm looping through the `UserSuggestions` array to create two sub-arrays. The first one has the users whose display names begin with a prefix provided by the user. The second one has the users whose display names don't begin with the prefix provided by the user. 

- I'm sorting both arrays alphabetically, then I combine the sorted arrays. 

- I updated the tests to conform with the new sorting criteria.

## Preview

<p float="left">
  <img src="https://user-images.githubusercontent.com/8026597/186367370-7a22106e-60f6-4e9e-933a-9ca93ffeb799.gif" width="300" />
</p>

## Testing Instructions

**N.B: The user-mentions suggestions feature is not enabled in self-hosted sites ( wordpress.org ).**

1. Tap the "My Site" tab.
2. Tap on "Menu"
3. Tap on "Comments" row.
4. Tap any comment.
5. Tap "Reply" button.
6. Type "@" and you should see the user-mentions suggestions list.
7. Type a prefix and check if the list is sorted first by users whose usernames or display names begin with the prefix, then alphabetically.

There are other scenarios to display the user-mentions suggestions list described in https://github.com/wordpress-mobile/WordPress-iOS/pull/18979.

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

9. What automated tests I added (or what prevented me from doing so)
I updated the existing tests related to the user-mentions suggestions list.

PR submission checklist:
